### PR TITLE
feat: support call kclvm native capi by gorpc

### PR DESCRIPTION
**Pull request by NeverRaR**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [x] N
- [ ] Y 

#### 2. What is the scope of this PR (e.g. component or file name):

pkg/3rdparty/dlopen/
gorpc
#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [x] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other
feat: support call kclvm native capi by gorpc

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

[gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go](https://github.com/KusionStack/kclvm-go/compare/main...NeverRaR:dev/c_api_call_client?expand=1#diff-874d909cb518cf0825e8912a6a1bd93d163a36f84188c78a77d3c1c2d6ab3b6b)
[gorpc/pkg/c_api_call/exec_program_test.go](https://github.com/KusionStack/kclvm-go/compare/main...NeverRaR:dev/c_api_call_client?expand=1#diff-ec273ed1bbea3831f4a6e5586a151fe93cd236b47094e92fdf27637a769647f0)

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: gorpc/pkg/c_api_call/cgo.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+}
+const char* kclvm_service_call(void *f,kclvm_service* c,const char * method,const char * args){
+	const char* (*service_call)(kclvm_service*,const char *,const char *);
+	service_call = (const char* (*)(kclvm_service*,const char *,const char *))f;
+	return service_call(c,method,args);
+}
+*/
+import "C"
+
+func NewKclvmService(pluginAgent C.longlong) *C.kclvm_service {
```
Comment (Peefy) [inline gorpc/pkg/c_api_call/cgo.go:37]: more documents on pub functions.

**Review by Peefy — DISMISSED** [review]
LGTM

File: gorpc/pkg/c_api_call/util.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	libSuffix := ".so"
+
+	if sysType == "darwin" {
+		libSuffix = ".dylib"
+	} else if sysType == "windows" {
+		libSuffix = ".dll"
+	}
+
+	libName := "libkclvm_capi" + libSuffix
+	if err == nil {
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/util.go]: the conditon `err == nil` always true

File: gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	out := &gpyrpc.Ping_Result{}
+	err = client.Ping(&gpyrpc.Ping_Args{Value: "hello"}, out)
+	assert.Nil(t, err)
+	assert.Equal(t, "hello", out.Value)
+}
+
+func TestExecProgram(t *testing.T) {
+	for j := 0; j < 10; j++ {
+		wg :=
+			sync.WaitGroup{}
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go]: fmt code, `var wg sync.WaitGroup`

File: gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+		n := 20
+		wg.Add(n)
+		for i := 0; i < n; i++ {
+			go func() {
+
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go]: remove blank line

File: gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+				workdir, _ := filepath.Abs(CORRECT_DATA_PATH)
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go]: check error

File: gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+						{Name: "__kcl_test_run", Value: "___test_schema_@@@__"},
+						{Name: "__kcl_test_debug", Value: "true"},
+					},
+					Overrides:         []*gpyrpc.CmdOverrideSpec{},
+					DisableYamlResult: false,
+				}
+				client := PROTOCAPI_NewKclvmServiceClient()
+				client.ExecProgram(args, &gpyrpc.ExecProgram_Result{})
+				println("ok")
+				//println(out.JsonResult)
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go]: remove debug code and comment

File: gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+					KFilenameList: []string{"complex.k"},
+					Args: []*gpyrpc.CmdArgSpec{
+						{Name: "__kcl_test_run", Value: "___test_schema_@@@__"},
+						{Name: "__kcl_test_debug", Value: "true"},
+					},
+					Overrides:         []*gpyrpc.CmdOverrideSpec{},
+					DisableYamlResult: false,
+				}
+				client := PROTOCAPI_NewKclvmServiceClient()
+				client.ExecProgram(args, &gpyrpc.ExecProgram_Result{})
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go]: check error

File: gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+				println("ok")
+				//println(out.JsonResult)
+				wg.Done()
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go]: use `defer wd.Done()` at the entry

File: gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	err = client.Ping(&gpyrpc.Ping_Args{Value: "hello"}, out)
+	assert.Nil(t, err)
+	assert.Equal(t, "hello", out.Value)
+}
+
+func TestExecProgram(t *testing.T) {
+	for j := 0; j < 10; j++ {
+		wg :=
+			sync.WaitGroup{}
+		n := 20
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/proto_capi_kclvm_service_test.go]: `const N = 20`

File: gorpc/pkg/c_api_call/proto_capi_kclvm_service_client.go [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+	c.client = NewKclvmService(C.longlong(kcl_plugin.GetInvokeJsonProxyPtr()))
+	runtime.SetFinalizer(c, func(x *PROTOCAPI_KclvmServiceClient) {
+		DeleteKclvmService(x.client)
+		x.client = nil
+	})
+	return c
+}
+
+func (c *PROTOCAPI_KclvmServiceClient) cApiCall(callName string, in proto.Message, out protoreflect.ProtoMessage) error {
+	type Validator interface {
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/proto_capi_kclvm_service_client.go:44]: should check args:
- `if callName == ""`
- if `if in == nil`
- if `if out == nil`

File: gorpc/pkg/c_api_call/proto_capi_kclvm_service_client.go [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+	goOut := C.GoString(cOut)
+
+	if strings.HasPrefix(goOut, "KCLVM_CAPI_CALL_ERROR") {
+		return errors.New(goOut)
+	}
+
+	return proto.Unmarshal([]byte(goOut), out)
+}
+
+func (c *PROTOCAPI_KclvmServiceClient) Ping(in *gpyrpc.Ping_Args, out *gpyrpc.Ping_Result) (err error) {
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/proto_capi_kclvm_service_client.go:91]: `if out == nil { ... }`, this `nil` is nil Pointer, not nil interface!!!

File: gorpc/pkg/c_api_call/exec_program_test.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+func TestExecErrorSingleFile(t *testing.T) {
+	client := PROTOCAPI_NewKclvmServiceClient()
+	files := getFiles(ERROR_DATA_PATH, ".k", true)
+	for _, file := range files {
+		_, err := exec(file, client)
+		assert.NotNil(t, err)
+	}
+}
+
+func exec(fileName string, client *PROTOCAPI_KclvmServiceClient) (out *gpyrpc.ExecProgram_Result, err error) {
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/exec_program_test.go]:
```go
func tExecProgram(t *testing.T, ...) {
    _, err := ...
    if err != nil {
        t.Fatal(err)
    }
}
```

File: gorpc/pkg/c_api_call/cgo.go [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+	freeString, _ := lib.GetSymbolPointer(f)
+
+	C.kclvm_service_free_string(freeString, str)
+}
+
+// KclvmServiceCall calls kclvm service by c api
+// args should be serialized as protobuf byte stream
+func KclvmServiceCall(serv *C.kclvm_service, method *C.char, args *C.char) *C.char {
+	f := "kclvm_service_call"
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/cgo.go]: const fnName = "..."

File: gorpc/pkg/c_api_call/cgo.go [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+	deleteServ, _ := lib.GetSymbolPointer(f)
+
+	C.kclvm_service_delete(deleteServ, serv)
+}
+
+// KclvmServiceFreeString releases the memory of  the return value of KclvmServiceCall
+func KclvmServiceFreeString(str *C.char) {
+	f := "kclvm_service_free_string"
+
+	freeString, _ := lib.GetSymbolPointer(f)
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/cgo.go]: check error

File: gorpc/pkg/c_api_call/cgo.go [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+	newServ, _ := lib.GetSymbolPointer(f)
+
+	return C.kclvm_service_new(newServ, pluginAgent)
+}
+
+// NewKclvmService releases the memory of kclvm_service
+func DeleteKclvmService(serv *C.kclvm_service) {
+	f := "kclvm_service_delete"
+
+	deleteServ, _ := lib.GetSymbolPointer(f)
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/cgo.go]: check error

File: gorpc/pkg/c_api_call/cgo.go [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+import "C"
+
+// NewKclvmService takes a pluginAgent  and returns  a pointer of capi kclvm_service.
+// pluginAgent is the address of C function pointer with type :const char * (*)(const char *method,const char *args,const char *kwargs),
+// in which "method" is the fully qualified name of plugin method,
+// and "args" ,"kwargs"  and return value of pluginAgent are JSON string
+func NewKclvmService(pluginAgent C.longlong) *C.kclvm_service {
+	f := "kclvm_service_new"
+
+	newServ, _ := lib.GetSymbolPointer(f)
```
Comment (chai2010) [inline gorpc/pkg/c_api_call/cgo.go]: check error

**Review by Peefy — APPROVED** [review]
LGTM

**Review by chai2010 — APPROVED** [review]
LGTM

**Pull request merged**